### PR TITLE
136 patients endpoint

### LINKED ISSUE #136: Programs document endpoint
Add endpoint that give hierarchical access to patient / program documents, e.g.

```graphql
patient {
   programs {
      encounters {
        ...
      }
   }
}
```

### PR DESCRIPTION
**Pull request by clemens-msupply**
- Adds StringFilter (currently needed for starts_width queries, not needed if using a separate patient -> doc table)
- Add document_registry node to the DocumentNode, e.g. to get the ui schema for the given document
- Add patients -> programs -> encounter nodes to the PatientNode
- Add single patient endpoint

Closes #136

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Can you put a database in the share folder so I can test?

With my current database I'm just getting "null" on the documentRegistry node for a patient that has a document.

E.g. Query:
```
query MyQuery {
  patient(storeId: "D77F67339BF8400886D009178F4962E1", patientId: "patient2") {
    id
    document {
      name
      type
      schema {
        jsonSchema
      }
      documentRegistry {
        id
        jsonSchema
        name
        uiSchema
      }
      data
    }
    name
  }
}
```

returns
```
{
  "data": {
    "patient": {
      "id": "patient2",
      "document": {
        "name": "patients/patient2",
        "type": "Patient",
        "schema": null,
        "documentRegistry": null,
        "data": {
          ...
          "firstName": "Peter",
          "gender": "Male",
          "healthCenter": {
            "id": "healthCenterId",
            "name": "My Health Center"
          },
          "id": "patient2",
          ...
      },
      "name": "Peter,Smith"
    }
  }
}
```

**Comment by clemens-msupply** [comment]
yes that is to be expected because the registry is empty (registry tells you how documents should be created but you can created your own documents as well, at least its not enforced that a document needs to have an entry in the registry)

will fill some dummy entries into the DB

File: server/graphql/core/src/loader/document_registry.rs [hunk 1]
```diff
@@ -6,6 +6,32 @@ use async_graphql::*;
...
+pub struct DocumentRegistryLoader {
+    pub service_provider: Data<ServiceProvider>,
+}
+
+#[async_trait::async_trait]
+impl Loader<String> for DocumentRegistryLoader {
+    type Value = DocumentRegistry;
+    type Error = RepositoryError;
+
+    async fn load(&self, entries: &[String]) -> Result<HashMap<String, Self::Value>, Self::Error> {
```
Comment (andreievg) [inline server/graphql/core/src/loader/document_registry.rs:18]: should this be document_types (vs entries)

File: server/graphql/document/src/queries/patient.rs [hunk 2]
```diff
@@ -99,6 +101,23 @@ impl PatientNode {
 
         Ok(result)
     }
+
+    pub async fn programs(&self, ctx: &Context<'_>) -> Result<Vec<ProgramNode>> {
+        let context = ctx.service_provider().context()?;
```
Comment (andreievg) [inline server/graphql/document/src/queries/patient.rs:106]: Although I don't think we will be querying for multiple patients including programs/documents I still think loader is more appropriate here. At least a TODO as you've done for encounters on program

File: server/repository/src/db_diesel/filter_sort_pagination.rs [hunk 3]
```diff
@@ -22,6 +22,103 @@ impl SimpleStringFilter {
     }
 }
 
+#[derive(Clone, PartialEq, Debug)]
+pub struct StringFilter {
```
Comment (andreievg) [inline server/repository/src/db_diesel/filter_sort_pagination.rs:26]: I think at this point, when we have three filters, it's easier on devs to consolidate them all into one. Not in this PR, but would be good to keep in mind

File: server/repository/src/diesel_macros.rs [hunk 4]
```diff
@@ -37,6 +37,69 @@ macro_rules! apply_equal_filter {
...
+                // Use case insensitive like
+                $query = $query.filter($dsl_field.ilike(format!("%{}%", value)));
+            }
+            if let Some(value) = filter.starts_with {
+                // Use case insensitive like
+                $query = $query.filter($dsl_field.ilike(format!("{}%", value)));
+            }
+            if let Some(value) = filter.ends_with {
+                // Use case insensitive like
+                $query = $query.filter($dsl_field.ilike(format!("%{}", value)));
```
Comment (andreievg) [inline server/repository/src/diesel_macros.rs:97]: hmm, any performance concerns with this filter ?

File: server/service/src/document/patient/query.rs [hunk 5]
```diff
@@ -85,6 +88,35 @@ pub fn get_patients(
...
+pub fn get_patient_program_encounters(
+    ctx: &ServiceContext,
+    store_id: &str,
+    patient_id: &str,
+    program: &str,
+) -> Result<Vec<Document>, RepositoryError> {
+    DocumentRepository::new(&ctx.connection).query(Some(
+        DocumentFilter::new()
+            .name(StringFilter::starts_with(
+                &patient_program_encounter_doc_name(patient_id, program, ""),
```
Comment (andreievg) [inline server/service/src/document/patient/query.rs:114]: I still think this is anti pattern for relational databases, and requires inspecting code to figure out how data is stored. I would like to revisit this sometime (and the requirements for it).

File: server/graphql/document/src/types/program.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /// The program document
+    pub async fn encounters(&self, ctx: &Context<'_>) -> Result<Vec<EncounterNode>> {
+        // TODO use loader
+        let context = ctx.service_provider().context()?;
+        let entries = ctx
+            .service_provider()
+            .patient_service
+            .get_patient_program_encounters(
```
Comment (andreievg) [inline server/graphql/document/src/types/program.rs:27]: If there are many encounters and many edits on those encounters, we would want to select the latest edit for each encounters (not all of the edits and all of the counters), is that already implemented ?

**Review by andreievg — APPROVED** [review]
Looks good, I think, some questions

File: server/graphql/core/src/loader/document_registry.rs (hunk 1 above)
Comment (clemens-msupply) [reply server/graphql/core/src/loader/document_registry.rs]: Thanks done. For filters I tend to use some generic name like `value` to help with copy and paste but this doesn't work very well for loaders. [9077397](https://github.com/openmsupply/open-msupply/pull/184/commits/907739759593a0a28e07adeebe21ca489a003e7c)

File: server/graphql/document/src/queries/patient.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/graphql/document/src/queries/patient.rs:106]: Yes agree I added the missing todo

File: server/repository/src/db_diesel/filter_sort_pagination.rs (hunk 3 above)
Comment (clemens-msupply) [reply server/repository/src/db_diesel/filter_sort_pagination.rs:26]: Yes at least StringFilter and SimpleStringFilter

File: server/repository/src/diesel_macros.rs (hunk 4 above)
Comment (clemens-msupply) [reply server/repository/src/diesel_macros.rs:97]: Probably results in a full column scan (?) I only added it for completeness. Our normal like filter does %{}% as well.

Think this filter might get handy for patients search we discussed today as well...

File: server/graphql/document/src/types/program.rs (hunk 6 above)
Comment (clemens-msupply) [reply server/graphql/document/src/types/program.rs:27]: You only get the latest encounter. But currently there is no pagination to limit how many encounters you get back. For example, if you have 1000 encounters you always get the latest version of these 1000 document.

File: server/service/src/document/patient/query.rs (hunk 5 above)
Comment (clemens-msupply) [reply server/service/src/document/patient/query.rs:114]: You mean patient_program_encounter_doc_name? we need a deterministic way to find a document name, e.g. Creating two HIVPrograms on two sites at the same sites should result in a single document

**Comment by clemens-msupply** [comment]
thanks for the review

**Pull request merged**